### PR TITLE
added documentation for Kernel#sprintf method (fixes #3124) 

### LINKED ISSUE #3124: #printf docs missing
The API doc of both `Kernel#printf` and `#sprintf` refer to `IO#printf` for documentation. But the two overloads there are actually [undocumented](https://github.com/crystal-lang/crystal/blob/68783f1ebdd5b9bbefc8892ddbc200fe1f13bbea/src/io.cr#L350-L358) (first being empty, second saying `ditto` 😆).

(Maybe unrelated, but in the source I see that `String::Formatter` is used, but for some reason that class is hidden from the API doc. Maybe that could be a good place to put the documentation?)

### PR DESCRIPTION
**Pull request by CaDs**
Issue: https://github.com/crystal-lang/crystal/issues/3124

As suggested per @asterite [on this previous PR](https://github.com/crystal-lang/crystal/pull/3126#discussion_r74092917)  I have used [Ruby's sprintf documentation](https://github.com/ruby/ruby/blob/96db72ce38b27799dd8e80ca00696e41234db6ba/sprintf.c) as a template.

### CODE REVIEW

File: src/kernel.cr [hunk 1]
```diff
@@ -68,8 +68,224 @@ def printf(format_string, args : Array | Tuple) : Nil
...
+# ```
+#
+# A format specifier consists of a percent sign, followed by optional flags,
+# width, and precision indicators, then terminated with a field type
+# character.  The field type controls how the corresponding
+# `sprintf` argument is to be interpreted, while the flags
+# modify that interpretation.
+#
+# The field type characters are:
+# ```pre
```
Comment (Sija) [inline src/kernel.cr]: What's `pre`?

Comment (straight-shoota) [reply src/kernel.cr]: `text` would probably be better.

Comment (CaDs) [reply src/kernel.cr]: changed to text

**Comment by r00ster91** [comment]
I would change the lines from 88 to 152 to:
```
# The field type characters are:
# ```text
#     Field │  Integer Format
#     ──────│──────────────────────────────────────────────────────────────
#       b   │ Formats argument as a binary number.
#       d   │ Formats argument as a decimal number.
#       i   │ Same as d.
#       o   │ Formats argument as an octal number.
#       x   │ Formats argument as a hexadecimal number using lowercase letters.
#       X   │ Same as x, but uses uppercase letters.
#
#     Field │  Float Format
#     ──────│──────────────────────────────────────────────────────────────
#       e   │ Formats floating point argument into exponential notation
#           │ with one digit before the decimal point as [─]d.dddddde[│─]dd.
#           │ The precision specifies the number of digits after the decimal
#           │ point (defaulting to six).
#       E   │ Equivalent to e, but uses an uppercase E to indicate
#           │ the exponent.
#       f   │ Formats floating point argument as [─]ddd.dddddd,
#           │ where the precision specifies the number of digits after
#           │ the decimal point.
#       g   │ Formats a floating point number using exponential form
#           │ if the exponent is less than ─4 or greater than or
#           │ equal to the precision, or in dd.dddd form otherwise.
#           │ The precision specifies the number of significant digits.
#       G   │ Equivalent to g, but use an uppercase E in exponent form.
#       a   │ Formats floating point argument as [─]0xh.hhhhp[│─]dd,
#           │ which is consisted from optional sign, "0x", fraction part
#           │ as hexadecimal, "p", and exponential part as decimal.
#       A   │ Equivalent to a, but use uppercase X and P.
#
#     Field │  Other Format
#     ──────│──────────────────────────────────────────────────────────────
#       c   │ Argument is the numeric code for a single character or
#           │ a single character string itself.
#       s   │ Argument is a string to be substituted.  If the format
#           │ sequence contains a precision, at most that many characters
#           │ will be copied.
#       %   │ A percent sign itself will be displayed.  No argument taken.
# ```
# The flags modifies the behavior of the formats.
# The flag characters are:
# ```text
#   Flag     │ Applies to    │ Meaning
#   ─────────│───────────────│─────────────────────────────────────────
#   space    │ bdiouxX       │ Leave a space at the start of
#            │ aAeEfgG       │ non─negative numbers.
#            │ (numeric fmt) │ For o, x, X, b, use
#            │               │ a minus sign with absolute value for
#            │               │ negative values.
#   ─────────│───────────────│─────────────────────────────────────────
#   +        │ bdiouxX       │ Add a leading plus sign to non─negative
#            │ aAeEfgG       │ numbers.
#            │ (numeric fmt) │ For o, x, X, b, use
#            │               │ a minus sign with absolute value for
#            │               │ negative values.
#   ─────────│───────────────│─────────────────────────────────────────
#   -        │ all           │ Left─justify the result of this conversion.
#   ─────────│───────────────│─────────────────────────────────────────
#   0 (zero) │ bdiouxX       │ Pad with zeros, not spaces.
#            │ aAeEfgG       │ For o, x, X, b, radix─1
#            │ (numeric fmt) │ is used for negative numbers formatted as
#            │               │ complements.
# ```
```
it uses │ and ─ instead of | and -

**Comment by straight-shoota** [comment]
@r00ster91 I don't think that'd be very useful. Using regular pipe and dash are totally fine. This way there will probably be less changes required once we can use Markdown table syntax.

File: src/kernel.cr [hunk 2]
```diff
@@ -68,8 +68,224 @@ def printf(format_string, args : Array | Tuple) : Nil
...
+#     Field |  Other Format
+#     ------+--------------------------------------------------------------
+#       c   | Argument is the numeric code for a single character or
+#           | a single character string itself.
+#       s   | Argument is a string to be substituted.  If the format
+#           | sequence contains a precision, at most that many characters
+#           | will be copied.
+#       %   | A percent sign itself will be displayed.  No argument taken.
+# ```
+# The flags modifies the behavior of the formats.
```
Comment (straight-shoota) [inline src/kernel.cr]: `modify`.

And please add newlines between code fences and paragraphs and between two paragraphs.

File: src/kernel.cr [hunk 3]
```diff
@@ -68,8 +68,224 @@ def printf(format_string, args : Array | Tuple) : Nil
...
+# The flag characters are:
+# ```text
+#   Flag     | Applies to    | Meaning
+#   ---------+---------------+-----------------------------------------
+#   space    | bdiouxX       | Leave a space at the start of
```
Comment (straight-shoota) [inline src/kernel.cr]: `Leave a space...` -> `Add a leading space character to non-negative numbers.` (use same wording as next entry)

File: src/kernel.cr [hunk 4]
```diff
@@ -68,8 +68,224 @@ def printf(format_string, args : Array | Tuple) : Nil
...
+# sprintf "%g", 123.4567       # => "123.457"
+# sprintf "%20.8g", 1234.56789 # => "           1234.5679"
+# sprintf "%20.8g", 123456789  # => "       1.2345679e+08"
+# sprintf "%20.8G", 123456789  # => "       1.2345679E+08"
+# sprintf "%20.8g", -123456789 # => "       -1.2345679e+08"
+# sprintf "%20.8G", -123456789 # => "       -1.2345679E+08"
+# ```
+#
+# The field width is an optional integer, followed optionally by a
+# period and a precision.  The width specifies the minimum number of
```
Comment (straight-shoota) [inline src/kernel.cr]: two spaces (ditto in the following paragraph)

File: src/kernel.cr [hunk 5]
```diff
@@ -68,8 +68,224 @@ def printf(format_string, args : Array | Tuple) : Nil
...
+#
+# For numeric fields, the precision controls the number of decimal places
+# displayed.  For string fields, the precision determines the maximum
+# number of characters to be copied from the string.  (Thus, the format
+# sequence <code>%10.10s</code> will always contribute exactly ten
+# characters to the result.)
+#
+# Examples of precisions:
+#
+# precision for `d`, `o`, `x` and `b` is
```
Comment (straight-shoota) [inline src/kernel.cr]: This sentence should be properly formatted (start with capital letter, period at the end). Ditto for the following.

**Comment by CaDs** [comment]
hmm build is failing, I guess I need to rebase master again?

**Comment by sdogruyol** [comment]
@CaDs please run `crystal tool format` and update

**Comment by CaDs** [comment]
Thanks @sdogruyol  🙇

**Comment by sdogruyol** [comment]
Thanks @CaDs :+1: Can you squash this?

**Comment by straight-shoota** [comment]
@sdogruyol it can be squashed when merged

**Comment by sdogruyol** [comment]
@straight-shoota yup :smile:

**Comment by bcardiff** [comment]
Thanks @CaDs !

**Pull request merged**